meetings/working-groups/collection-literals/CL-2024-01-23.md

```diff
@@ -0,0 +1,143 @@
+
+2. The natural type of a collection expression with some contextual *element-target-typing*.  This would apply to cases like so:
+
+    - `foreach (byte b in [1, 2, 3])`.  Here, we believe that the `byte` type should help influence the collection type being created (so that this case is legal).
+
+    - `List<Predicate> list = [a, b, .. c ? [_ => true] : []]`.  Similarly, the use of `Predicate` here in the target should help influence the type of `[_ => true]` such that this code succeeds.
+
+Overall though, this space is enormous and will need a lot of future design.  Open areas include, but are not limited to:
+
+1. Could we envision a world where `var v = [];` is ever allowed?  This would involve flow analysis to determine the element type based on how `v` was used.
+
+2. Should we align on a well known type for the natural type, or should we consider it just a special language type (like anonymous-types) that the compiler makes its own determinations about.  For example, we could pick a well known type like `List<T>` or we could consider the language having a special ``builtin-list<T>`` type with special properties.
+
+3. Should we try to mandate a very efficient type (like Span/InlineArrays) for perf?  How would that work with async/await?
+
+4. Would it be possible to pick an inefficient type (like `List<T>`) but then give the compiler broad leeway to optimize in the common case where it would not be observable (for example, actually emitting on the stack when safe).
+
+### Extension methods
+
+We are still interesting in enabling `[a, b, c].ExtensionMethod()`. However, we think this is not something that should be limited to collection expressions.  Broadly speaking, the language is very limited wrt to the interaction of extension methods and target-typed constructs.  For example, you cannot do this `(x => true).ExtensionOnStringPredicate()`.  The language requires the item being invoked have a type *prior* to lookup of the extension method.  To make cases like this work, we'd need to allow them to not have a type, and then say that lookup should still find the extensions, which are then tested for applicability in the rewritten form (e.g. `ExtensionMethod([a, b, c])`).
+
+As we are investing heavily in extensions in C#13, we may want to roll the exploration of this area into that work.
+
+### Addressing existing types that should work with collection expressions, but don't.
+
+There are several types we've found that annoying don't work with collection expressions.  For example:
+
+- ArraySegment<T>
+- ReadOnlyMemory<T>.  Commonly used as the analog of ReadOnlySpan<T> when you're working with async methose.
```

Comment:
Need backticks to prevent `<T>` parsing as a markdown HTML element.

```suggestion
- `ArraySegment<T>`
- `ReadOnlyMemory<T>`.  Commonly used as the analog of `ReadOnlySpan<T>` when you're working with async methods.
```